c/planarityTestGraphFunctionality.c

```diff
@@ -140,36 +160,100 @@ int TestGraphFunctionality(char *commandString, char *infileName, char *inputStr
 					char *finalSlash = strrchr(infileName, FILE_DELIMITER);
 					char *infileBasename = finalSlash ? (finalSlash + 1) : infileName;
 
-					char *headerStr = (char *) malloc((strlen(infileBasename) + 3) * sizeof(char));
+					char *headerFormat = "FILENAME=\"%s\" DURATION=\"%.3lf\"\n";
+					char *headerStr = (char *) malloc(
+														(
+															strlen(headerFormat) +
+															strlen(infileBasename) +
+															strlen("-1.7976931348623158e+308") + // -DBL_MAX from float.h
+															3
+														) * sizeof(char));
 					if (headerStr == NULL)
 					{
 						ErrorMessage("Unable allocate memory for output file header.\n");
+
 						free(inputStr);
 						inputStr = NULL;
+
 						gp_Free(&theGraph);
+
 						sf_Free(&testOutput);
+
 						return NOTOK;
 					}
 
-					sprintf(headerStr, "%s\n", infileBasename);
+					testAllStatsP stats = (testAllStatsP) calloc(1, sizeof(testAllStats));
+					if (stats == NULL)
+					{
+						ErrorMessage("Unable allocate memory for test stats.\n");
+
+						free(inputStr);
+						inputStr = NULL;
+
+						gp_Free(&theGraph);
+
+						sf_Free(&testOutput);
+
+						free(headerStr);
+						headerStr = NULL;
+
+						return NOTOK;
+					}
+
+					char command = commandString[1];
+					Result = testAllGraphs(theGraph, command, inputStr, stats);
+
+					// Stop the timer
+					platform_GetTime(end);
+					duration = platform_GetDuration(start, end);
+					sprintf(messageContents, "\nDone testing all graphs (%.3lf seconds).\n", duration);
+					Message(messageContents);
+
+					sprintf(headerStr, headerFormat, infileBasename, duration);
 					sf_fputs(headerStr, testOutput);
 					free(headerStr);
 					headerStr = NULL;
 
-					Result = testAllGraphs(theGraph, commandString[1], inputStr, testOutput);
-
-					if (Result != OK)
+					char *resultsStr = (char *) malloc(
+														(
+															3 +_getNumCharsToReprInt(stats->numGraphsRead) +
+															1 + _getNumCharsToReprInt(stats->numOK) +
+															1 + _getNumCharsToReprInt(stats->numNONEMBEDDABLE)+
+															1 + 8 + // either ERROR or SUCCESS, so the longer of which is 7 + 1 chars
+															3
+														) * sizeof(char));
+					if (resultsStr == NULL)
 					{
-						errorMessageFormatStr = "Unable to perform algorithm corresponding to command '%c' to graph(s).\n";
-						sprintf(messageContents, errorMessageFormatStr, commandString[1]);
-						ErrorMessage(messageContents);
+						ErrorMessage("Unable allocate memory for results string.\n");
+
+						free(inputStr);
+						inputStr = NULL;
+
+						free(stats);
+						stats = NULL;
+
+						gp_Free(&theGraph);
+
+						sf_Free(&testOutput);
+
+						return NOTOK;
```

Comment:
You have return NOTOK here, but below at the end there are some return -1 instances that seem like they should be return NOTOK as well?

Reply:
I think you're right, that all instances of `return -1` in `TestGraphFunctionality` should be changed to `return NOTOK`, since:
* `commandLine` of `planarityCommandLine.c` (which calls `callTestGraphFunctionality`, which calls `TestGraphFunctionality`) expects `OK`, `NONEMBEDDABLE`, or `NOTOK` from all the other functions it calls
* `runGraphTransformationTest` handles the conversion from `OK`, `NONEMBEDDABLE`, and `NOTOK` by `return (Result == OK) ? 0 : -1;` except in one case: when `ConstructTransformationExpectedResultFilename` doesn't return `OK` or `expectedOutfileName` is still `NULL`, I changed `return NOTOK` to `return -1`
  * This is because `runQuickRegressionTests` of `planarityCommandLine.c` calls `runSpecificGraphTests` and returns `NOTOK` if its return value is `< 0`; `runSpecificGraphTests` calls `runGraphTransformationTest`, and if its return value is `< 0`, sets `retVal` to `-1`
* `TransformMenu` and `TestMenu` of `planarity.c` expects `OK`, `NONEMBEDDABLE`, or `NOTOK`